openedx/core/lib/request_utils.py

```diff
@@ -1,16 +1,49 @@
 """ Utility functions related to HTTP requests """
 import re
+from urlparse import urlparse
+import crum
 
 from django.conf import settings
+from django.test.client import RequestFactory
 from opaque_keys import InvalidKeyError
 from opaque_keys.edx.keys import CourseKey
-
 from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
 
+
 # accommodates course api urls, excluding any course api routes that do not fall under v*/courses, such as v1/blocks.
 COURSE_REGEX = re.compile(r'^(.*?/courses/)(?!v[0-9]+/[^/]+){}'.format(settings.COURSE_ID_PATTERN))
 
 
+def get_request_or_stub():
+    """
+    Return the current request or a stub request.
+
+    If called outside the context of a request, construct a fake
+    request that can be used to build an absolute URI.
+
+    This is useful in cases where we need to pass in a request object
+    but don't have an active request (for example, in tests, celery tasks, and XBlocks).
+    """
+    request = crum.get_current_request()
+
+    if request is None:
+
+        # The settings SITE_NAME may contain a port number, so we need to
+        # parse the full URL.
+        full_url = "http://{site_name}".format(site_name=settings.SITE_NAME)
```

Comment:
Is there any concern about the "http" portion of this (as opposed to getting it from settings somehow)?

Reply:
We don't use the "http" portion of the temporary URL that is created here.  It looks like this code creates a temporary URL just so it can then parse it to get the hostname and port.  It is weird, though, that production code is using `RequestFactory` from the test framework.

Reply:
Yeah. I can't place my finger on exactly how, but this feels pretty sketchy from the security point of view. But since this was existing code that was moved here, I'm not going to block on it.